From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Val=C3=A9rio=20Val=C3=A9rio?= <valerio.valerio@jolla.com>
Date: Mon, 23 Feb 2015 13:31:30 +0200
Subject: [PATCH] Prevent push enabled status to go out of sync.

---
 .../messageservices/imap/imapservice.cpp      | 55 +++++++++++--------
 1 file changed, 32 insertions(+), 23 deletions(-)

diff --git a/src/plugins/messageservices/imap/imapservice.cpp b/src/plugins/messageservices/imap/imapservice.cpp
index 9aa53472..e4f08297 100644
--- a/src/plugins/messageservices/imap/imapservice.cpp
+++ b/src/plugins/messageservices/imap/imapservice.cpp
@@ -64,7 +64,10 @@ QString connectionSettings(ImapConfiguration &config)
     result << config.mailServer();
     result << QString::number(config.mailPort());
     result << QString::number(config.mailEncryption());
+    // When accounts-qt is used auth is automaticly set from the advertised capabilities
+#ifndef USE_ACCOUNTS_QT
     result << QString::number(config.mailAuthentication());
+#endif
     return result.join(QChar('\x0A')); // 0x0A is not a valid character in any connection setting
 }
 
@@ -1479,6 +1482,11 @@ ImapService::ImapService(const QMailAccountId &accountId)
       _networkSession(0),
       _networkSessionTimer(new QTimer(this))
 {
+#ifdef USE_KEEPALIVE
+    _accountPushEnabled = false;
+    _buteoReplyReceived = false;
+#endif
+
     QMailAccount account(accountId);
     if (account.status() & QMailAccount::Enabled) {
         enable();
@@ -1495,8 +1503,6 @@ ImapService::ImapService(const QMailAccountId &accountId)
 
     // Connect to dbus signals emitted by buteo notifying schedule changes
     QDBusConnection m_dBusConnection(QDBusConnection::sessionBus());
-    _accountPushEnabled = false;
-    _buteoReplyReceived = false;
 
     if(!m_dBusConnection.isConnected()) {
         qWarning() << Q_FUNC_INFO << "Cannot connect to Dbus";
@@ -1525,22 +1531,19 @@ void ImapService::enable()
     ImapConfiguration imapCfg(accountCfg);
     bool pushEnabled = accountPushEnabled();
 #ifdef USE_KEEPALIVE
-    _client->setPushEnabled(_accountPushEnabled);
     // When account is enabled or account state changes, request push status from buteo scheduler
     if (imapCfg.pushCapable() && !_buteoReplyReceived) {
-        _accountWasPushEnabled = false;
+        _accountPushEnabled = false;
         pushEnabled = false;
         QDBusMessage message = QDBusMessage::createMethodCall("com.meego.msyncd", "/synchronizer", "com.meego.msyncd",
                                                               "isSyncedExternally");
         uint acct = _accountId.toULongLong();
         message.setArguments(QVariantList() << acct << "syncemail");
         QDBusConnection::sessionBus().asyncCall(message);
-    } else {
-        _accountWasPushEnabled = _accountPushEnabled;
     }
-#else
-    _accountWasPushEnabled = pushEnabled;
+    _client->setPushEnabled(_accountPushEnabled);
 #endif
+    _accountWasPushEnabled = pushEnabled;
     _previousPushFolders = imapCfg.pushFolders();
     _previousConnectionSettings = connectionSettings(imapCfg);
     
@@ -1603,6 +1606,7 @@ void ImapService::accountsUpdated(const QMailAccountIdList &ids)
     if (_client) {
         loggingIn = _client->loggingIn();
     }
+
     if (!isEnabled) {
         if (_accountWasEnabled) {
             // Account changed from enabled to disabled
@@ -1710,7 +1714,7 @@ void ImapService::initiatePushEmail()
 #ifdef USE_KEEPALIVE
     QMailAccountConfiguration accountCfg(_accountId);
     ImapConfiguration imapCfg(accountCfg);
-    _accountWasPushEnabled = _accountPushEnabled;
+    _accountWasPushEnabled = accountPushEnabled();
     _previousPushFolders = imapCfg.pushFolders();
 #endif
     QMailFolderIdList ids(_client->configurationIdleFolderIds());
@@ -1960,22 +1964,27 @@ void ImapService::pushEnabledStatus(uint accountId, const QString &profileType,
 {
     qMailLog(Messaging) << "Received new idleState for account: " << accountId << "state: " << state << "profile type: " << profileType;
     if (accountId == _accountId.toULongLong() && profileType == QLatin1String("syncemail")) {
-        _buteoReplyReceived = true;
-        if (state != _accountPushEnabled) {
-            qMailLog(Messaging) << Q_FUNC_INFO << "Changing push enabled state to: " << state;
-            _accountPushEnabled = state;
-            if (_accountPushEnabled) {
-                if (_accountWasEnabled) {
-                    disable();
+        QMailAccount account(_accountId);
+        if (account.status() & QMailAccount::Enabled) {
+            _buteoReplyReceived = true;
+            if (state != _accountPushEnabled) {
+                qMailLog(Messaging) << Q_FUNC_INFO << "Changing push enabled state to: " << state;
+                _accountPushEnabled = state;
+                if (_accountPushEnabled) {
+                    if (_accountWasEnabled) {
+                        disable();
+                    }
+                    enable();
+                } else {
+                    // When it changes to true, this state is set inside "enable()"
+                    // since a new client is created
+                    _client->closeIdleConnections();
+                    _client->setPushEnabled(state);
+                    stopPushEmail();
                 }
-                enable();
-            } else {
-                // When it changes to true, this state is set inside "enable()"
-                // since a new client is created
-                _client->closeIdleConnections();
-                _client->setPushEnabled(state);
-                stopPushEmail();
             }
+        } else {
+            qMailLog(Messaging) << Q_FUNC_INFO << "Account" << accountId << " is disabled, not changing push enabled state";
         }
     }
 }
