From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Damien Caliste <dcaliste@free.fr>
Date: Wed, 19 May 2021 09:23:33 +0200
Subject: [PATCH] Revert loadRelax()

---
 src/libraries/qmfclient/qprivateimplementation.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/libraries/qmfclient/qprivateimplementation.h b/src/libraries/qmfclient/qprivateimplementation.h
index 3f5e1daf..21c76d1b 100644
--- a/src/libraries/qmfclient/qprivateimplementation.h
+++ b/src/libraries/qmfclient/qprivateimplementation.h
@@ -99,7 +99,7 @@ public:
 
     inline int count() const
     {
-        return ref_count.loadRelaxed();
+        return ref_count.load();
     }
 
     inline void ref()
